### PR TITLE
Add Auto Quantize in ModelOpt quantize example

### PR DESCRIPTION
**Pull request by jenchen13**
# What does this PR do ?
- support auto quantize to ModelOpt quantize example
- See https://github.com/NVIDIA/Model-Optimizer/pull/1562 for changes made in Model-Optimizer to support AutoQuantize in Megatron

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

# Test run

On Nemotron Nano 3, running MMLU via mmlu.sh
BF16: 0.705 MMLU, 513/728
AutoQuantize at 4.8 effective bits, calib seq len 8192, calib size 512, autoquant score size 256: 0.702 MMLU, 511/728
calib dataset: CNN dailymail

Retains 99.6% of BF16 score.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: examples/post_training/modelopt/quantize.py [hunk 1]
```diff
@@ -29,6 +32,7 @@
 # releases.
 try:
     from modelopt.torch.utils.plugins.megatron_calibration import (
+        get_megatron_calibration_dataloader,
```
Comment (kevalmorabia97) [inline examples/post_training/modelopt/quantize.py:35]: is this newly added?

File: examples/post_training/modelopt/quantize.py [hunk 2]
```diff
@@ -161,6 +173,45 @@ def add_text_generate_ptq_args(parser):
...
+        help="Method for auto_quantize sensitivity scoring.",
+    )
+    group.add_argument(
+        "--auto-quantize-score-size",
+        type=int,
+        default=128,
+        help="Number of samples to use for sensitivity scoring in auto_quantize.",
+    )
+    group.add_argument(
+        "--auto-quantize-checkpoint",
```
Comment (ChenhanYu) [inline examples/post_training/modelopt/quantize.py:207]: Is this checkpoint save/restore aware of TP/EP/PP sharding? Theoretically, per rank will store a checkpoint?

File: examples/post_training/modelopt/quantize.py [hunk 3]
```diff
@@ -312,6 +405,70 @@ def get_calib_dataloader(
...
+        forward_backward_step=None,
+        disabled_layers=disabled_layers,
+        num_calib_steps=num_calib_steps,
+        num_score_steps=num_score_steps,
+        verbose=True,
+        method=args.auto_quantize_method,
+        checkpoint=args.auto_quantize_checkpoint,
+    )
+
+    if args.save is not None and torch.distributed.get_rank() == 0:
```
Comment (ChenhanYu) [inline examples/post_training/modelopt/quantize.py:478]: I am a bit confused here. Why save the search_state here? Isn't that it should have been save above?

File: examples/post_training/modelopt/quantize.py [hunk 4]
```diff
@@ -312,6 +405,70 @@ def get_calib_dataloader(
...
+        data_loader=calib_dataloader,
+        forward_step=forward_step,
+        loss_func=None,
+        forward_backward_step=None,
+        disabled_layers=disabled_layers,
+        num_calib_steps=num_calib_steps,
+        num_score_steps=num_score_steps,
+        verbose=True,
+        method=args.auto_quantize_method,
+        checkpoint=args.auto_quantize_checkpoint,
```
Comment (ChenhanYu) [inline examples/post_training/modelopt/quantize.py:475]: I assume this is for both save and restore which is useful for slurm timeout? Does it also do per rank `.pth` as below?

**Comment by ChenhanYu** [comment]
Does auto-quantized checkpoint restored directly for export (which only support PP) or QAD?

File: examples/post_training/modelopt/quantize.py (hunk 1 above)
Comment (jenchen13) [reply examples/post_training/modelopt/quantize.py:35]: yes, see the Model-Opt PR https://github.com/NVIDIA/Model-Optimizer/pull/1562
It was carved out from megatron_calibration to be reused here since the dataloader is needed by `auto_quantize`

File: examples/post_training/modelopt/quantize.py (hunk 2 above)
Comment (jenchen13) [reply examples/post_training/modelopt/quantize.py:207]: it might not be, TBH it is not recommended to save/restore the auto quant search state in MLM. I haven't tested that enough, I should remove this

Comment (jenchen13) [reply examples/post_training/modelopt/quantize.py:207]: according to Cursor it is aware of TP/EP/DP, but rejects PP (PP is not compatible with AutoQuant)

```
Yes — one file per global rank. With torch.distributed initialized, each rank writes/loads its own file (rank0.pth, rank1.pth, or checkpoint0.pth, checkpoint1.pth, depending on whether the path is a directory or file).
```

File: examples/post_training/modelopt/quantize.py (hunk 4 above)
Comment (jenchen13) [reply examples/post_training/modelopt/quantize.py:475]: `--auto-quantize-checkpoint` is for intermediate resuming AutoQuant in case the job crashes. `--save` saves a rank 0 only save of the final search state, used to inspect the scores per layer after the job finishes

**Comment by jenchen13** [comment]
@ChenhanYu yes you can use the auto quantize checkpoint for export or QAD. it is just a PTQ checkpoint

**Review by kevalmorabia97 — APPROVED** [review]
LGTM

**Comment by jenchen13** [comment]
/ok to test [0429f9f](https://github.com/NVIDIA/Megatron-LM/pull/4821/commits/0429f9fe6243c92e494895012f7078d7d3d4e351)
